KAFKA-6386:use Properties instead of StreamsConfig in KafkaStreams constructor

abbccdda opened the pull request:
This pull request targets https://issues.apache.org/jira/browse/KAFKA-6386
The minor fix to deprecate usage of `StreamsConfig` in favor of `java.util.Properties`.
I created separate public constructors using `Properties` in order to replace the old ones, 
and prioritize new functions in the `KafkaStreams.java` file.

Since this is my first time doing open source contribution, I'm very happy to get
any comment or pointer to be more professional and get better next time, thank you Guozhang @guozhangwang and Liquan @Ishiihara!

testing strategy: existing unit test should be suffice to cover this change.

Commit:
use Properties instead of StreamsConfig in KafkaStreams constructor

abbccdda commented:
@mjsax Hey Matt, could u help take a look? Thanks!

mjsax commented:
Thanks for the PR. I had a quick look an it's looks quite good. However, you need to do a KIP as mentioned in the JIRA before we can merge this. Please read the KIP webpage that explains how this works -- if you have any question about the KIP process, let us know.

abbccdda commented:
Sounds good, thanks! @mjsax

abbccdda commented:
@mjsax Hey Matt, just done as requirement, feel free to let me know your thoughts!  https://cwiki.apache.org/confluence/display/KAFKA/KIP-245%3A+Use+Properties+instead+of+StreamsConfig+in+KafkaStreams+constructor

mjsax commented:
Thanks for the KIP @abbccdda . Looks good. You need to start a DISCUSS thread on the dev mailing list and a VOTE thread in a second step (as describe in the KIP page). For this simple KIP, you could skip the DISCUSS thread and start a VOTE immediately, too.

abbccdda commented:
@mjsax thanks for the guidance. I just sent out a VOTE thread to mailing list. Let me know your thoughts :)

abbccdda commented:
@guozhangwang @mjsax Per request on Guozhang's comment in vote, I added a fourth constructor which takes in both `KafkaClientSupplier` and `Time`. Let me know your thoughts!

abbccdda commented:
Hey @mjsax, do you think the updated diff looks good to you?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -819,12 +819,12 @@ public synchronized final TopologyBuilder copartitionSources(final Collection<St
     /**
      * Build the topology for the specified topic group. This is called automatically when passing this builder into the
-     * {@link org.apache.kafka.streams.KafkaStreams#KafkaStreams(TopologyBuilder, org.apache.kafka.streams.StreamsConfig)} constructor.
+     * {@link org.apache.kafka.streams.KafkaStreams#KafkaStreams(TopologyBuilder, java.util.Properties)} constructor.
```
We don't need to update `TopologyBuilder` because it's a deprecated class.

mjsax reviewed (commented):
Just one nit.

Does your dev mailing list setup work now?

guozhangwang commented:
@abbccdda are you still working on this PR?

mjsax commented:
@abbccdda Today is feature freeze for 1.1 release. If we can't merge this by end of the day, we have to defer it for 1.2 release.

abbccdda commented:
@mjsax @guozhangwang so sorry for the delay! I will update the diff over the weekend and have another review.

abbccdda commented:
@mjsax addressed your previous comments and rebase. Sorry again for the delay!

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/318/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/319/

mjsax commented on streams/src/main/java/org/apache/kafka/streams/KafkaClientSupplier.java:
```diff
@@ -26,7 +26,7 @@
  * {@code KafkaClientSupplier} can be used to provide custom Kafka clients to a {@link KafkaStreams} instance.
  *
- * @see KafkaStreams#KafkaStreams(org.apache.kafka.streams.Topology, StreamsConfig, KafkaClientSupplier)
+ * @see KafkaStreams#KafkaStreams(org.apache.kafka.streams.Topology, java.util.Properties, KafkaClientSupplier)
```
nit: `java.util.` can be removed (note, we only specify the whole package for `Topology`, because otherwise we would need to add an `import` statement and get a warning about "unused imports" and a failing build.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -123,7 +123,7 @@
  *
  *
- * @see KafkaStreams#KafkaStreams(org.apache.kafka.streams.Topology, StreamsConfig)
+ * @see KafkaStreams#KafkaStreams(org.apache.kafka.streams.Topology, java.util.Properties)
```
nit `java.util.` can be removed

mjsax reviewed (commented):
Thanks for updating the PR (and no need to apologize for anything).

Two more nits.

Did you check the output for `./gradlew clean jar testJar` for warnings about deprecation usage of `KafkaStreams` constructors? You did fix some already, but I was wondering if those are all?

mjsax commented:
What is the status of this PR @abbccdda ?

mjsax commented:
@abbccdda Any updates for this PR?

mjsax commented:
@abbccdda Just wanted to follow up here. Any updates?

abbccdda commented:
@mjsax So sorry I was overwhelmed by my work these days. I should be able to make some progress by the end of next week. Thanks for the follow up!

abbccdda commented:
Addressed the review comments. I'm debugging on why the build is failed due to 400+ deprecation warnings.

mjsax commented:
@abbccdda 

I think this is the problem:
```
06:24:45 :streams:checkstyleMain[ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk7-scala2.11/
streams/src/main/java/org/apache/kafka/streams/KafkaClientSupplier.java:25:8: Unused import - java.util.Properties. [UnusedImports]
06:24:47  FAILED
06:24:47 
06:24:47 FAILURE: Build failed with an exception.
06:24:47 
06:24:47 * What went wrong:
06:24:47 Execution failed for task ':streams:checkstyleMain'.
```

The deprecation warning should not be an issue. Also note, that many of those are fixed already in `trunk` but your PR is open for some time already and thus does not have the latest changes from `trunk`. You could rebase your PR if you want -- but it's not required.

abbccdda commented:
@mjsax Thanks for taking a look. I just found the same root cause tonight, and change back the comment to `@see KafkaStreams#KafkaStreams(Topology, java.util.Properties, KafkaClientSupplier)` like before. The only thing I'm confused was the previous comment you gave me that `java.util.Properties` is not good to use in comment, where I suppose is ok since it is only a javadoc reference? 
I bet there are still some deprecated usage in the codebase since my PR opens quite long now... I shall revisit the code again tomorrow and fix the new usage if possible. Thank you!!!

mjsax commented:
I see. Seems that checkstyle is not able to pick up the usage of the import with JavaDocs `@see` tag...

mjsax reviewed (approved):
Thanks for updating this PR! LGTM.

Call for second review @bbejeck @vvcephei @guozhangwang

guozhangwang commented:
Code change LGTM. Could we update the `docs/streams/upgrade-guide.html` in `Streams API changes in 1.2.0` section to mention that this constructor has been deprecated?

abbccdda commented:
@mjsax @guozhangwang thanks so much for the revisit! So the doc change Guozhang mentioned is also in the codebase right? I could do a final update tonight.

guozhangwang commented:
> @mjsax @guozhangwang thanks so much for the revisit! So the doc change Guozhang mentioned is also in the codebase right? I could do a final update tonight.

Yup, it is in the same repo, under `docs` folder.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -579,25 +571,70 @@ public KafkaStreams(final Topology topology,
+     * @deprecated use {@link #KafkaStreams(Topology, Properties, KafkaClientSupplier)} instead
+     */
+    @Deprecated
+    public KafkaStreams(final org.apache.kafka.streams.processor.TopologyBuilder builder,
```
here and above, do we need full package names?

mjsax replied:
It's intentional to avoid build warnings about importing deprecated classes.

bbejeck reviewed (approved):
Just one minor question, otherwise LGTM

mjsax commented:
The doc changes should go into `docs/streams/upgrade-guide.html`.

abbccdda commented:
Been waiting the build for a while... Anyway, addressed the review comments again. @mjsax @guozhangwang @bbejeck Let me know if there are more things you feel needed to change. Thank you all!

mjsax commented on docs/streams/upgrade-guide.html:
```diff
@@ -86,7 +86,11 @@ <h3><a id="streams_api_changes_120" href="#streams_api_changes_120">Streams API
-    
+
+     <p>
+	We have deprecated <code>StreamsConfig</code> in <code>KafkaStreams</code> constructors. Now we only take in <code>java.util.Properties</code> since <code>StreamsConfig</code> is immutable and is created from a Properties object itself.
```
Nit: avoid "tabs" and use whitespace only.

Sounds a little bit as if `StreamsConfig` was deprecates. What about:
```
We deprecated all constructors of <code>KafkaStreams</code> that take a <code>StreamsConfig</code> as parameter.
Please use the corresponding constructors that accept <code>java.util.Properties</code> instead.
For more details, see <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-245%3A+Use+Properties+instead+of+StreamsConfig+in+KafkaStreams+constructor">KIP-245</a>.
```

I don't think we need to explain why we changed it (people can read the KIP if they are interested).

mjsax reviewed (commented):
One minor comment about the docs. Overall LGTM.

guozhangwang commented:
I have incorporated @mjsax 's comment while merging to trunk. Thanks @abbccdda !

guozhangwang reviewed (commented):
LGTM.

abbccdda commented:
@mjsax @guozhangwang Thank you so much for helping me through the process! Really appreciate your patience on my slow progress in the past month. I will make sure to make daily progress next time if I'm working on a PR.

mjsax commented:
Thanks for contributing! It's not easy if you do this "on the side" instead of full time. Really appreciate your work! No need to apologize.

(When we ping on regular basis, it's not to put pressure on anybody, but to see if the PR was abandoned or not only.)
